File: slack-reader/src/test/java/dev/number6/slackreader/core/SlackReaderIntegrationTest.java
package dev.number6.slackreader.core;

import com.amazonaws.services.lambda.runtime.LambdaLogger;
import com.google.gson.Gson;
import dev.number6.db.model.ChannelComprehensionSummary;
import dev.number6.message.ChannelMessages;
import dev.number6.slackreader.SlackReader;
import dev.number6.slackreader.dagger.*;
import org.junit.jupiter.api.Test;

import java.time.Clock;
import java.time.LocalDate;
import java.util.*;
import java.util.stream.Collectors;

import static org.assertj.core.api.Assertions.assertThat;

class SlackReaderIntegrationTest {

    private final Gson gson = new Gson();
    private final LambdaLogger logger = new LambdaLogger() {
        @Override
        public void log(String message) {
            System.out.println(message);
        }

        @Override
        public void log(byte[] message) {
            System.out.println(Arrays.toString(message));
        }
    };

    private final TestSlackReaderComponent testee = DaggerTestSlackReaderComponent.create();

    @Test
    void handleEventWithoutDate() {

        Clock clock = testee.getClock();
        Map<String, Object> triggerEvent = new HashMap<>();
        testEventComprehensionDate(triggerEvent, LocalDate.now(clock));
    }

    @Test
    void handleEventWithDate() {
        Clock clock = testee.getClock();
        Map<String, Object> triggerEvent = new HashMap<>();
        triggerEvent.put(SlackReader.COMPREHENSION_DATE_FIELD_NAME, LocalDate.now(clock).toString());
        testEventComprehensionDate(triggerEvent, LocalDate.now(clock));
    }

    private void testEventComprehensionDate(Map<String, Object> triggerEvent, LocalDate comprehensionDate) {

        RecordingSlackReaderAdaptor reader = (RecordingSlackReaderAdaptor) testee.getSlackPort();
        FakeDynamoDbMapperModule.FakeAmazonDynamoDB fakeDynamo = testee.getFakeAmazonDynamoClient();
        FakeAmazonSnsModule.FakeAmazonSns fakeSns = testee.getFakeAmazonSns();

        if (triggerEvent == null) {
            triggerEvent = new HashMap<>();
        } else {
            triggerEvent.put(SlackReader.COMPREHENSION_DATE_FIELD_NAME, comprehensionDate.toString());
        }
        testee.handler().handle(triggerEvent, logger);

        List<ChannelComprehensionSummary> savedChannelSummaries = fakeDynamo.getSavedObjects().stream()
                .map(o -> ((ChannelComprehensionSummary) o))
                .collect(Collectors.toList());

        savedChannelSummaries.forEach(s -> assertThat(s.getComprehensionDate()).isEqualTo(comprehensionDate));
        assertThat(savedChannelSummaries.stream().map(ChannelComprehensionSummary::getChannelName).collect(Collectors.toList()))
                .containsExactlyInAnyOrderElementsOf(reader.getChannelNames());

        Collection<ChannelMessages> channelMessages = fakeSns.getPublishedMessages().stream()
                .map(p -> gson.fromJson(p, ChannelMessages.class))
                .collect(Collectors.toList());

        channelMessages.forEach(cm -> assertThat(cm.getChannelName()).isIn(reader.getChannelNames()));
        channelMessages.forEach(cm -> assertThat(cm.getMessages()).containsExactlyInAnyOrderElementsOf(reader.getMessagesForChannelName(cm.getChannelName())));
    }
}

File: slack-reader/src/test/kotlin/dev/number6/slackreader/core/SlackReaderIntegrationTest.kt
package dev.number6.slackreader.core

import assertk.assertThat
import assertk.assertions.containsOnly
import assertk.assertions.isEqualTo
import assertk.assertions.isIn
import com.amazonaws.services.lambda.runtime.LambdaLogger
import com.google.gson.Gson
import dev.number6.db.model.ChannelComprehensionSummary
import dev.number6.message.ChannelMessages
import dev.number6.slackreader.SlackReader
import dev.number6.slackreader.dagger.DaggerTestSlackReaderComponent
import dev.number6.slackreader.dagger.RecordingSlackReaderAdaptor
import org.junit.jupiter.api.Test
import java.time.LocalDate
import java.util.*

internal class SlackReaderIntegrationTest {
    private val gson = Gson()
    private val logger: LambdaLogger = object : LambdaLogger {
        override fun log(message: String?) {
            println(message)
        }

        override fun log(message: ByteArray?) {
            println(Arrays.toString(message))
        }
    }
    private val testee = DaggerTestSlackReaderComponent.create()

    @Test
    fun handleEventWithoutDate() {
        val clock = testee.getClock()
        val triggerEvent: Map<String, Any> = mapOf()
        testEventComprehensionDate(triggerEvent, LocalDate.now(clock))
    }

    @Test
    fun handleEventWithDate() {
        val clock = testee.getClock()
        val triggerEvent: Map<String, Any> = mapOf(SlackReader.COMPREHENSION_DATE_FIELD_NAME to LocalDate.now(clock).toString())
        testEventComprehensionDate(triggerEvent, LocalDate.now(clock))
    }

    private fun testEventComprehensionDate(event: Map<String, Any>?, comprehensionDate: LocalDate?) {
        var triggerEvent = event
        val reader = testee.getSlackReaderPort() as RecordingSlackReaderAdaptor
        val fakeDynamo = testee.getFakeAmazonDynamoClient()
        val fakeSns = testee.getFakeAmazonSns()
        triggerEvent = triggerEvent?.plus(SlackReader.COMPREHENSION_DATE_FIELD_NAME to comprehensionDate.toString())
                ?: mapOf()
        testee.handler().handle(triggerEvent, logger)
        val savedChannelSummaries = fakeDynamo.getSavedObjects().map { o -> o as ChannelComprehensionSummary }

        savedChannelSummaries.forEach { s -> assertThat(s.comprehensionDate).isEqualTo(comprehensionDate) }

        val savedNames: List<String> = savedChannelSummaries.map { ccs -> ccs.channelName }
        val readerNames = reader.getChannelNames().toTypedArray()
        assertThat(savedNames).containsOnly(*readerNames)

        val channelMessages = fakeSns.published
                .map { p: String? -> gson.fromJson(p, ChannelMessages::class.java) }

        channelMessages.forEach { cm -> assertThat(cm.channelName).isIn(*readerNames) }
        channelMessages.forEach { cm -> assertThat(cm.messages).containsOnly(*reader.getMessagesForChannelName(cm.channelName).toTypedArray()) }
    }
}